# AI4EAC Liquidity Stress Prediction: End-to-End Champion Pipeline

This notebook executes the complete 5-stage leak-free tournament reproduction and ensemble stacking pipeline.

### Pipeline Configuration:
- **Cross-Validation**: 1-Seed (), 10-Fold Stratified CV across all stages.
- **Validation Integrity**: Strictly nested feature screening, zero validation fold leakage, and genuine out-of-sample calibration.

### Stage Architecture:
1. **Stage 1**: Baseline Family Anchor (1-Seed 10-Fold CatBoost, XGBoost, LightGBM).
2. **Stage 2**: 1-Seed 10-Fold Domain GBDT Zoo (Nested feature screening + resilient fold metadata across 5 architectures).
3. **Stage 3**: Dual TabPFN Foundation Priors (Physics & Solvency ratio representations).
4. **Stage 4**: Diversity Base Learners (MultiStrata Stratified Ensemble + De-risked Distillation Students + Neural TabMLP).
5. **Stage 5**: 10-Fold Cross-Validated Logit-Space L2 Regularized Meta-Stacker + Single-Stage Cross-Fitted Calibration.
6. **Stage 6**: Automated Competition Invariant Audit (30k rows, probability clipping, zero NaNs, natural prevalence).

In [ ]:
import os
import sys
import time
import numpy as np
import pandas as pd

# Ensure repo root is on sys.path
sys.path.insert(0, os.path.abspath("."))

from src.config import SEED, N_SPLITS, get_hardware_info, get_default_dataset_paths

print(f"✓ Environment Initialized | Global Seed: {SEED} | Folds: {N_SPLITS}")
print(f"✓ Hardware Info: {get_hardware_info()}")
tr_p, te_p = get_default_dataset_paths()
print(f"✓ Dataset Paths: Train={tr_p}, Test={te_p}")


## Stage 1: Tournament Baseline Anchor Reproduction & Hill Climbing
Trains 4-seed CatBoost, XGBoost, and HistGradientBoosting on the 87 selected features, optimizes weights via SLSQP, applies Beta/Platt calibration, and runs stepwise hill climbing.

In [ ]:
from src.stages.stage1_reproduce import run_stage1
import time
t0 = time.time()
s1_score = run_stage1()
print(f"✓ Stage 1 Complete in {(time.time() - t0)/60:.2f} min | Composite Score: {s1_score:.5f}")

STAGE 1: COMPLETE BASELINE + HILL CLIMBING PIPELINE

--- [Stage 1A] Baseline 1-Seed Unified Family Ensemble (10-fold CV) ---
Step 1: Engineering comprehensive feature set
Loaded cached features from checkpoints/feature_cache (906 features)
Built multi-strata cohort keys (39 unique strata) for balanced CV splitting
Step 2: Running feature_engine selection (targeting ~60 features)
--- Feature Selection: Initial count = 904 features ---
feature_engine.DropConstantFeatures: dropped 7 features
feature_engine.DropDuplicateFeatures: dropped 145 features
feature_engine.SmartCorrelatedSelection (r>0.98, unsupervised variance): dropped 62 features
--- Feature Selection Complete: 79 features selected (out of 904) ---
Step 3: Multi-seed cross-validation — 10-fold CV, seeds=[42], features=79 (74 numeric + 5 categorical)

--- Seed 1/1 : seed=42 ---
  [seed=42] Training model: catboost
    Fold 1 | LogLoss: 0.2318 | AUC: 0.9198 | Comp: 0.7341
    Fold 2 | LogLoss: 0.2411 | AUC: 0.9154 | Comp: 0.7230


  Fold 1/10 | Comp: 0.73464 | AUC: 0.92036 | LL: 0.23156


  Fold 2/10 | Comp: 0.72274 | AUC: 0.91496 | LL: 0.24122


  Fold 3/10 | Comp: 0.71529 | AUC: 0.90788 | LL: 0.24579
  Fold 4/10 | Comp: 0.72892 | AUC: 0.91639 | LL: 0.23565
  Fold 5/10 | Comp: 0.74618 | AUC: 0.92873 | LL: 0.22343
  Fold 6/10 | Comp: 0.72331 | AUC: 0.91445 | LL: 0.24045
  Fold 7/10 | Comp: 0.74822 | AUC: 0.92594 | LL: 0.22030
  Fold 8/10 | Comp: 0.74297 | AUC: 0.92398 | LL: 0.22474


## Stage 2: 1-Seed 10-Fold Multi-Model Domain GBDT Zoo
Executes CatBoost, XGBoost, and LightGBM with leak-free fold-nested feature screening across specialized domain subsets (solvency ratios, balance collapses, and triage columns).

In [1]:
from src.stages.stage2_gbdt_zoo import run_stage2
import time
t0 = time.time()
s2_score = run_stage2(seeds=(42, ), n_splits=10)
print(f"✓ Stage 2 Complete in {(time.time() - t0)/60:.2f} min | Composite Score: {s2_score:.5f}")

STAGE 2: 1-SEED 10-FOLD DOMAIN GBDT ZOO (5 ARCHITECTURES)
✓ Loaded Stage Champion Anchor: Comp=0.73409 | AUC=0.91968 | LL=0.23184

Step 1: Engineering comprehensive features & filtering via feature_engine
Loaded cached features from checkpoints/feature_cache (906 features)
Built multi-strata cohort keys (39 unique strata) for balanced CV splitting
--- Feature Selection: Initial count = 904 features ---
feature_engine.DropConstantFeatures: dropped 7 features
feature_engine.DropDuplicateFeatures: dropped 145 features
feature_engine.SmartCorrelatedSelection (r>0.98, unsupervised variance): dropped 62 features
--- Feature Selection Complete: 79 features selected (out of 904) ---
Active features for Domain Zoo: 79 (5 categoricals)
Domain Views Configured:
  dom2_cols (Physics + Digital Channels)  : 29 features
  dom3_cols (Physics + Digital + Momentum): 36 features
  dom_triage_cols (Physics + Triage)      : 14 features

[1/5] Training Domain Architecture: cb_d6_dom26 (cols=29)
  ==> cb_d6_

## Stage 3: Dual TabPFN Foundation Priors
Extracts non-tree foundation model representations on GPU using RankGauss Quantile Transformation on the physical solvency ratio views.

In [ ]:
import time
from src.stages.stage3_tabpfn_priors import run_stage3

t0 = time.time()
s3_score = run_stage3(n_splits=10)
print(f"✓ Stage 3 Complete in {(time.time() - t0)/60:.2f} min | Composite Score: {s3_score:.5f}")

Stage 3: Dual TabPFN Foundation Priors
Loaded cached features from checkpoints/feature_cache (906 features)
Successfully authenticated TabPFN via Kaggle secret TABPFN_TOKEN.
Executing Genuine TabPFN Foundation Priors on GPU across 2 sub-views...
[pfn_phys] Features (14): ['agg_inflow_trend_ratio', 'agg_daily_avg_bal_recent3_to_old3_ratio', 'agg_withdraw_total_value_recent3_to_old3_ratio', 'agg_inflow_to_outflow_6m', 'agg_paybill_total_value_recent3_to_old3_ratio', 'agg_recent3_net_cashflow', 'agg_recent3_inflow', 'agg_withdraw_highest_amount_recent3_to_old3_ratio', 'x_90_d_activity_rate', 'agg_withdraw_volume_recent3_to_old3_ratio', 'agg_merchantpay_total_value_recent3_to_old3_ratio', 'agg_daily_avg_bal_cv', 'agg_daily_avg_bal_recent3_minus_old3', 'agg_net_cashflow_delta']


tabpfn-v3.5-20260909.safetensors:   0%|          | 0.00/876M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/36.0 [00:00<?, ?B/s]

  [pfn_phys] Fold 1/10 | LL: 0.2312 | AUC: 0.9202 | Comp: 0.7349 (466.3s)


## Stage 4: Diversity Pipeline (MultiStrata, Distillation Students, Neural TabMLP)
Executes:
1. MultiStrata 5-Fold Iterative Stratification Ensemble (906 features).
2. 10-Fold Self-Training Distillation Students (CatBoost & XGBoost) on 70k soft pseudo-targets.
3. PyTorch Neural TabMLP with residual connections and bounded logit scaling.

In [ ]:
import time
from src.stages.stage4_diversity import run_stage4

t0 = time.time()
s4_score = run_stage4()
print(f"✓ Stage 4 Complete in {(time.time() - t0)/60:.2f} min | Composite Score: {s4_score:.5f}")

## Stage 5: 10-Fold Cross-Validated Logit-Space L2 Meta-Stacker
Solves analytical L-BFGS-B regularized regression in logit space across all candidate model streams, eliminates in-sample hill-climbing, and applies single-stage cross-fitted Platt calibration.

In [ ]:
import time
from src.stages.stage5_meta_stacker import run_stage5

t0 = time.time()
final_champion_score = run_stage5()
print(f"★ STAGE 5 META-STACKER COMPLETE: Champion Score = {final_champion_score:.5f} ★")

## Stage 6: Final Verification & Invariants Audit
Checks all 5 hard competition invariants: exact row counts (30,000), column names, zero NaNs, probability bounding [0.002, 0.998], and benchmark correlation.

In [ ]:
import os
import pandas as pd
from src.stages.audit import run_audit

# Run automated audit
audit_passed = run_audit()
print(f"Audit Passed: {audit_passed}")

# Inspect final submission preview
sub_file = "submissions/submission_stage5_final.csv"
if os.path.exists(sub_file):
    df_sub = pd.read_csv(sub_file)
    print(f"Submission Shape: {df_sub.shape}")
    print(f"Mean Predicted Probability: {df_sub['Target'].mean():.5f}")
    print(f"5th Percentile: {df_sub['Target'].quantile(0.05):.5f} | 95th Percentile: {df_sub['Target'].quantile(0.95):.5f}")
    print("First 10 Rows:")
    print(df_sub.head(10))
else:
    print(f"{sub_file} not found yet.")